storage/src/index/storage.rs

```diff
@@ -0,0 +1,166 @@
+//! A memory efficient index structure for a journal that maps keys to locations.
+
+use crate::index::Translator;
+use commonware_runtime::Metrics;
+use prometheus_client::metrics::counter::Counter;
+use std::collections::{hash_map::Entry, HashMap};
+
+/// Each key is mapped to a `Record` that contains a potential location of the requested value in
+/// the journal, and an optional link to another potential location to allow for collisions.
+///
+/// The `Index` uses a `Translator` to transform keys into a compressed representation, resulting in
+/// non-negligible probability of collisions. Collision resolution is the responsibility of the
+/// user.
+struct Record {
+    location: u64,
+
+    next: Option<Box<Record>>,
+}
+
+pub struct LocationIterator<'a> {
+    next: Option<&'a Record>,
+}
+
+impl LocationIterator<'_> {
+    /// Create a `LocationIterator` that returns no items.
+    fn empty() -> Self {
+        LocationIterator { next: None }
+    }
+}
+
+impl Iterator for LocationIterator<'_> {
+    type Item = u64;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.next {
+            Some(next) => {
+                let loc = next.location;
+                self.next = next.next.as_deref();
+                Some(loc)
+            }
+            None => None,
+        }
+    }
+}
+
+impl Record {
+    fn iter(&self) -> LocationIterator {
+        LocationIterator { next: Some(self) }
+    }
+}
+
+pub struct Index<T: Translator, E: Metrics> {
+    translator: T,
+
+    // We store the first location associated with a key within the HashMap entry itself to
```

Comment:
nit: I think this comment is better suited for `Archive`? I'm not sure there is a heap anymore either.

Reply:
ok, removed. I thought this was referring to the Boxed entries.. the first u64 location is not boxed (in the map entry) whereas others are (& on the heap as a result).